cryptoki/tests/wycheproof.rs

```diff
@@ -0,0 +1,168 @@
+#[serial]
+fn aes_gcm_wycheproof() -> TestResult {
+    let (pkcs11, slot) = init_pins();
+    let session = pkcs11.open_rw_session(slot)?;
+    session.login(UserType::User, Some(&AuthPin::new(USER_PIN.into())))?;
+
+    // Load Wycheproof AES-GCM test vectors
+    let test_set = wycheproof::aead::TestSet::load(wycheproof::aead::TestName::AesGcm)?;
+
+    let mut passed = 0;
+    let mut failed = 0;
+    let mut skipped = 0;
+
+    for test_group in &test_set.test_groups {
+        let key_size = test_group.key_size;
+
+        // Only test key sizes we support (128, 192, 256 bits)
+        if ![128, 192, 256].contains(&key_size) {
+            skipped += test_group.tests.len();
+            continue;
+        }
+
+        for test in &test_group.tests {
+            // Skip tests with nonce sizes that exceed PKCS#11 limits (max 256 bytes)
+            if test.nonce.len() > 256 {
+                skipped += 1;
+                continue;
+            }
+
+            // Skip tests with tag sizes that exceed PKCS#11 limits (max 128 bits)
+            if test.tag.len() * 8 > 128 {
+                skipped += 1;
+                continue;
+            }
+            // Import the test key
+            let key_template = vec![
+                Attribute::Class(cryptoki::object::ObjectClass::SECRET_KEY),
+                Attribute::KeyType(cryptoki::object::KeyType::AES),
+                Attribute::Token(false),
+                Attribute::Sensitive(false),
+                Attribute::Extractable(true),
+                Attribute::Encrypt(true),
+                Attribute::Decrypt(true),
+                Attribute::Value(test.key.to_vec()),
+            ];
+
+            let key = match session.create_object(&key_template) {
+                Ok(k) => k,
+                Err(e) => {
+                    eprintln!("Test {}: Failed to create key: {:?}", test.tc_id, e);
+                    failed += 1;
+                    continue;
+                }
+            };
+
+            // Prepare GCM parameters (need mutable nonce for GcmParams)
+            let mut nonce = test.nonce.to_vec();
+            let tag_bits = match (test.tag.len() * 8).try_into() {
+                Ok(bits) => bits,
+                Err(e) => {
+                    eprintln!("Test {}: Failed to convert tag length: {:?}", test.tc_id, e);
+                    failed += 1;
+                    continue;
+                }
+            };
+            let gcm_params =
+                match cryptoki::mechanism::aead::GcmParams::new(&mut nonce, &test.aad, tag_bits) {
+                    Ok(params) => params,
+                    Err(e) => {
+                        eprintln!("Test {}: Failed to create GCM params: {:?}", test.tc_id, e);
+                        failed += 1;
+                        continue;
+                    }
+                };
+
+            // Test encryption
+            let encrypt_result = session.encrypt(&Mechanism::AesGcm(gcm_params), key, &test.pt);
```

Comment:
Could we test also the PKCS#11 3 API with message-based encryption?

https://github.com/parallaxsecond/rust-cryptoki/blob/36fb3c3a2a0dc9c625c61329322cd0cabbb9d853/cryptoki/tests/basic.rs#L2587